service/history/task/task.go

```diff
@@ -447,6 +454,21 @@ func (t *taskImpl) shouldResubmitOnNack() bool {
 		(t.queueType == QueueTypeActiveTransfer || t.queueType == QueueTypeActiveTimer)
 }
 
+func (t *taskImpl) shouldDropTaskOnError() bool {
```

Comment:
This helper is a little bit dangerous without the err being validated inside. I'd pass the error as an arg and check it, so the function can't be misused in the future.